vm_trace.c

```diff
@@ -726,7 +726,8 @@ call_trace_func(rb_event_flag_t event, VALUE proc, VALUE self, ID id, VALUE klas
         event != RUBY_EVENT_C_CALL &&
         event != RUBY_EVENT_C_RETURN &&
         (VM_FRAME_RUBYFRAME_P(ec->cfp) ||
-         VM_FRAME_RUBYFRAME_P(RUBY_VM_PREVIOUS_CONTROL_FRAME(ec->cfp)))) {
+         VM_FRAME_RUBYFRAME_P(RUBY_VM_PREVIOUS_CONTROL_FRAME(ec->cfp))) &&
+        imemo_type_p((VALUE)ec->cfp->iseq, imemo_iseq)) {
```

Comment:
> From my understanding this part checks whether `rb_vm_make_binding()` would raise before using it; it was checking for the opposite of the raise conditions in rb_vm_make_binding() before the assertion fix. Shouldn't the imemo check also go into `rb_vm_make_binding()` to keep the two consistent?

`rb_vm_make_binding()` doesn't raise in this case, it will happily create a binding for an ifunc cfp, which later causes the VM assertion failure during garbage collection. I think checking in `rb_vm_make_binding()` is too late, unless we are OK changing the method contract to allow it to return `nil` or raise instead of a returning binding if called for a ifunc frame.

> Also, it looks like this callback only ever wants to check the top most frame: [XrXr@0307c18](https://github.com/XrXr/ruby/commit/0307c18b03f38781c13156c3559213058eab356b)
> 
> Further, the crash from the added test is coming from inspecting the second top most frame in addition to the top most frame. The following even smaller diff is able to fix the assertion:
> 
> ```diff
> @@ -725,8 +725,7 @@ call_trace_func(rb_event_flag_t event, VALUE proc, VALUE self, ID id, VALUE klas
>      if (self && (filename != Qnil) &&
>          event != RUBY_EVENT_C_CALL &&
>          event != RUBY_EVENT_C_RETURN &&
> -        (VM_FRAME_RUBYFRAME_P(ec->cfp) ||
> -         VM_FRAME_RUBYFRAME_P(RUBY_VM_PREVIOUS_CONTROL_FRAME(ec->cfp)))) {
> +        VM_FRAME_RUBYFRAME_P(ec->cfp)) {
>          argv[4] = rb_binding_new();
>      }
>      argv[5] = klass ? klass : Qnil;
> ```

I believe I added checking the previous control frame to keep the tests working, but maybe there were additional changes that make it no longer required.  I'm happy to merge 0307c18b03f38781c13156c3559213058eab356b if it passes CI.

> Other code seem to assume that `VM_FRAME_RUBYFRAME_P(cfp)` implies `imemo_type_p(cfp->iseq, imemo_iseq)`, for example in `vm_make_env_each()`:
> 
> ```
>     if (!VM_FRAME_RUBYFRAME_P(cfp)) {
> 	local_size = VM_ENV_DATA_SIZE;
>     }
>     else {
>         local_size = ISEQ_BODY(cfp->iseq)->local_table_size + VM_ENV_DATA_SIZE;
>     }
> ```
> 
> Though that code might have its own problem unrelated to this patch set.

`VM_FRAME_RUBYFRAME_P(cfp)` is `!VM_FRAME_CFRAME_P(cfp)`, so it should still return true for ifuncs.  Not sure if that matters for ``vm_make_env_each`.

Maybe `rb_vm_get_binding_creatable_next_cfp` should return 0 instead of a cfp is the cfp is for an ifunc?  Currently, it only checks for `cfp->iseq`, it doesn't also check `imemo_type_p(cfp->iseq, imemo_iseq)`.

Reply:
>  Not sure if that matters for ``vm_make_env_each`

To answer my own question about the RUBYFRAME implication, it's probably because ifunc frames are always also `VM_FRAME_FLAG_CFRAME`. The two `VM_FRAME_MAGIC_IFUNC` usages also set the CFRAM flag.

> Maybe `rb_vm_get_binding_creatable_next_cfp` should return 0 instead of a cfp is the cfp is for an ifunc?

Yeah, that would also allow for some refactoring of the condition in `rb_vm_make_binding()` I think. It strikes me as a bit odd that the frames that are checked are not the frames used to make the binding later on, and the relationship between `src_cfp` and `cfp` is not so clear to me. I tried making that change locally but had some trouble because `rb_vm_get_binding_creatable_next_cfp()` has other users.

Anyways, that change is probably off topic for this patch set and I can try for a refactor later if desired.